core/src/main/java/io/grpc/internal/MessageDeframer.java

```diff
@@ -438,6 +449,18 @@ public synchronized void reset() throws IOException {
       count = mark;
     }
 
+    @Override
+    public void close() throws IOException {
+      super.close();
+    }
+
+    private void reportCount() {
+      if (count > maxCount) {
+        statsTraceContext.messageReceived(count - maxCount);
```

Comment:
Is messageReceived supposed to be compressed size, or uncompressed size? If compressed size, then it _can't_ correspond to high-level messages (e.g., protos). If the event is supposed to be called _before_ application-level flow control (you want to report when a message was actually received off the wire), then it _can't_ be uncompressed size. For uncompressed size, it can only be known immediately before calling the application.

> Alternatively, I could do this in close(), but that requires all Marshallers to close the InputStream. No Marshaller does it today. This would be an API change. Not that it would break users, but that stats won't work correctly if the implementation doesn't follow the new API.

That may interfere with zero-copy protobuf, which would retain buffers until close(), but close() needs to be delayed to allow the application to process. That isn't really designed, though. We should maybe speak with Nathan to see what ideas he had for it.

Instead of relying on close(), you could just notice EOF/-1 as returned by read() (and you could still handle close() as well, although in that case you may know that the message was not fully read.)